### PR TITLE
JBPM-6099: Navigation from process to tasks and vice versa

### PR DESCRIPTION
**Pull request by cristianonicolai**

### CODE REVIEW

**Comment by cristianonicolai** [comment]
Replacing https://github.com/kiegroup/jbpm-wb/pull/756

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/instance/list/ProcessInstanceListPresenter.java [hunk 1]
```diff
@@ -601,11 +606,20 @@ public void setupDefaultActiveSearchFilters() {
...
+                              pid);
+    }
+
+    public void openTaskAdminView(final String pid) {
+        navigateToPerspective(TASKS_ADMIN,
+                              SEARCH_PARAMETER_PROCESS_INSTANCE_ID,
+                              pid);
+    }
+
+    public boolean canOpenTaskAdminView() {
```
Comment (nmirasch) [inline jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/instance/list/ProcessInstanceListPresenter.java]: @cristianonicolai Do we have to verify the authorization to a view before each horizontal navigation to a view to cover the situations created by the authorization manager?

Comment (bsremac) [inline jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/instance/list/ProcessInstanceListPresenter.java]: Would recommend clearer method name here; along the lines of _isAuthorizedForTaskAdminView_

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/instance/list/ProcessInstanceListPresenter.java [hunk 2]
```diff
@@ -601,11 +606,20 @@ public void setupDefaultActiveSearchFilters() {
...
+        navigateToPerspective(JOBS,
+                              SEARCH_PARAMETER_PROCESS_INSTANCE_ID,
+                              pid);
+    }
+
+    public void openTaskAdminView(final String pid) {
+        navigateToPerspective(TASKS_ADMIN,
+                              SEARCH_PARAMETER_PROCESS_INSTANCE_ID,
+                              pid);
+    }
```
Comment (bsremac) [inline jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/instance/list/ProcessInstanceListPresenter.java]: Where is the method for navigation to user task perspective?

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/instance/list/ProcessInstanceListPresenter.java (hunk 1 above)
Comment (cristianonicolai) [inline jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/instance/list/ProcessInstanceListPresenter.java]: @nmirasch good point. I created a shared method to check the user permission to view certain perspective.
@bsremac removed this method, now is done via `getViewTasksActionCondition`

File: jbpm-wb-human-tasks/jbpm-wb-human-tasks-client/src/test/java/org/jbpm/workbench/ht/client/editors/taskslist/AbstractTaskListViewTest.java [hunk 3]
```diff
@@ -288,6 +284,21 @@ public void testSuspendActionHasCell(){
...
     
+    @Test
+    public void testViewProcessActionHasCell(){
+        ConditionalActionHasCell viewProcCell = new ConditionalActionHasCell(
+                "",
+                cellDelegate,
+                task -> (task.getProcessInstanceId() >= 0)
+        );
+        
+        TaskSummary testTask = createTestTaskSummary(TASK_STATUS_READY, TEST_USER_ID, POSITIVE_POTENTIAL_OWNERS);
```
Comment (cristianonicolai) [inline jbpm-wb-human-tasks/jbpm-wb-human-tasks-client/src/test/java/org/jbpm/workbench/ht/client/editors/taskslist/AbstractTaskListViewTest.java]: @bsremac status should not be a problem as the data still available even if process/task is completed/aborted. The only condition is to have access to the perspective.

File: jbpm-wb-human-tasks/jbpm-wb-human-tasks-client/src/test/java/org/jbpm/workbench/ht/client/editors/taskslist/AbstractTaskListViewTest.java [hunk 4]
```diff
@@ -288,6 +284,21 @@ public void testSuspendActionHasCell(){
...
+        runActionHasCellTest(testTask, viewProcCell, true);
+        testTask = new TaskSummary(0L, "", "", "", 0, "", "", new Date(), new Date(),
+                new Date(), "", 0, -1 /*This is processInstanceId*/, "", 0, new Date(), "", "");
+        runActionHasCellTest(testTask, viewProcCell, false);
```
Comment (cristianonicolai) [inline jbpm-wb-human-tasks/jbpm-wb-human-tasks-client/src/test/java/org/jbpm/workbench/ht/client/editors/taskslist/AbstractTaskListViewTest.java]: @bsremac moved tests to presenter, hope is better now.

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/instance/list/ProcessInstanceListPresenter.java (hunk 2 above)
Comment (cristianonicolai) [inline jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/instance/list/ProcessInstanceListPresenter.java]: @bsremac added now. the rule is first try to navigate to tasks admin if the user has access to, othwerwhise the regular task view is used.

**Comment by bsremac** [comment]
@cristianonicolai It seems you have been still making some changes to this PR.
I'm getting Refresh notifications as I'm going though it :)
Let me know when you finish and I will have one more look.

**Comment by cristianonicolai** [comment]
@bsremac sorry about that, I was fixing some issues after the last rebase and I notice some false positive tests as well. Should be good for review now.

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/test/java/org/jbpm/workbench/pr/client/editors/instance/list/ProcessInstanceListPresenterTest.java [hunk 5]
```diff
@@ -504,4 +641,79 @@ public void testActiveSearchFiltersProcessDefinitionId(){
...
+        final ArgumentCaptor<ResourceRef> captor = ArgumentCaptor.forClass(ResourceRef.class);
+        verify(authorizationManager).authorize(captor.capture(),
+                                               eq(identity));
+        assertEquals(TASKS_ADMIN,
+                     captor.getValue().getIdentifier());
+        assertEquals(ActivityResourceType.PERSPECTIVE,
+                     captor.getValue().getResourceType());
+
+        assertFalse(presenter.isUserAuthorizedForPerspective(TASKS_ADMIN));
+    }
```
Comment (bsremac) [inline jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/test/java/org/jbpm/workbench/pr/client/editors/instance/list/ProcessInstanceListPresenterTest.java]: Could we add one more unit test for _openTaskView_ method or re-factor this one to also include arguments validation of _navigateToPerspective_ method?
In this way, I think branching would be properly tested and I don't need to add Selenium test which does that.

File: jbpm-wb-human-tasks/jbpm-wb-human-tasks-client/src/main/java/org/jbpm/workbench/ht/client/editors/taskslist/AbstractTaskListPresenter.java [hunk 6]
```diff
@@ -583,4 +581,25 @@ public FilterSettings getVariablesTableSettings( String taskName ) {
...
+    }
+
+    protected Predicate<TaskSummary> getClaimActionCondition() {
+        return task -> task.getStatus().equals(TASK_STATUS_READY);
+    }
+
+    protected Predicate<TaskSummary> getReleaseActionCondition() {
+        return task -> task.getActualOwner() != null && task.getActualOwner().equals(identity.getIdentifier())
+                && (task.getStatus().equals(TASK_STATUS_RESERVED) || task.getStatus().equals(TASK_STATUS_IN_PROGRESS));
+    }
```
Comment (cristianonicolai) [inline jbpm-wb-human-tasks/jbpm-wb-human-tasks-client/src/main/java/org/jbpm/workbench/ht/client/editors/taskslist/AbstractTaskListPresenter.java]: @bsremac correct, that should fix JBPM-6022 and JBPM-6021

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/test/java/org/jbpm/workbench/pr/client/editors/instance/list/ProcessInstanceListPresenterTest.java (hunk 5 above)
Comment (cristianonicolai) [inline jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/test/java/org/jbpm/workbench/pr/client/editors/instance/list/ProcessInstanceListPresenterTest.java]: @bsremac sure, will do.

Comment (cristianonicolai) [inline jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/test/java/org/jbpm/workbench/pr/client/editors/instance/list/ProcessInstanceListPresenterTest.java]: @bsremac done

File: jbpm-wb-human-tasks/jbpm-wb-human-tasks-client/src/main/java/org/jbpm/workbench/ht/client/editors/taskslist/AbstractTaskListView.java [hunk 7]
```diff
@@ -229,26 +225,31 @@ private void onMouseOverGrid( ExtendedPagedTable<TaskSummary> extendedPagedTable
...
-        cells.add(new CompleteActionHasCell(constants.Open(), task -> {
-            presenter.selectTask(task, false);
-        }));
+        cells.add(new ConditionalButtonActionCell<TaskSummary>(constants.Resume(),
+                                                               task -> presenter.resumeTask(task),
+                                                               presenter.getResumeActionCondition()));
+
+        cells.add(new ConditionalButtonActionCell<TaskSummary>(constants.ViewProcess(),
+                                                               task -> presenter.openProcessInstanceView(task.getProcessInstanceId().toString()),
+                                                               presenter.getPrcessInstanceCondition()));
```
Comment (nmirasch) [inline jbpm-wb-human-tasks/jbpm-wb-human-tasks-client/src/main/java/org/jbpm/workbench/ht/client/editors/taskslist/AbstractTaskListView.java]: @cristianonicolai getProcessInstanceCondition :)

**Comment by cristianonicolai** [comment]
@nmirasch @bsremac thanks, changes done

**Pull request merged**